### PR TITLE
Add APIs to add client capabilities and claims

### PR DESCRIPTION
**Pull request by neha-bhargava**
Add APIs to the managed identity app and request to add client capabilities and claims. Bypass the cache when claims are passed.

### CODE REVIEW

File: msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/ManagedIdentityApplication.java [hunk 1]
```diff
@@ -99,9 +102,21 @@ public Builder resource(String resource) {
             return self();
         }
 
+        /**
+         * Sets client capabilities for the application.
+         * 
+         * @param clientCapabilities List of client capabilities to be requested
```
Comment (gladjohn) [inline msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/ManagedIdentityApplication.java]:
```suggestion
         * @param clientCapabilities a list of capabilities (e.g., ["cp1"]) recognized by the token service
```

File: msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/ManagedIdentityApplication.java [hunk 2]
```diff
@@ -99,9 +102,21 @@ public Builder resource(String resource) {
...
+         * @return instance of Builder of ManagedIdentityApplication
```
Comment (gladjohn) [inline msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/ManagedIdentityApplication.java]:
```suggestion
         * @return this Builder instance
```

File: msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/ManagedIdentityApplication.java [hunk 3]
```diff
@@ -99,9 +102,21 @@ public Builder resource(String resource) {
             return self();
         }
 
+        /**
+         * Sets client capabilities for the application.
```
Comment (gladjohn) [inline msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/ManagedIdentityApplication.java]:
```suggestion
         * Sets the client capabilities requested during authentication.
```

File: msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/ManagedIdentityParameters.java [hunk 4]
```diff
@@ -83,8 +86,13 @@ public ManagedIdentityParametersBuilder forceRefresh(boolean forceRefresh) {
             return this;
         }
 
+        public ManagedIdentityParametersBuilder claims(String claims) {
```
Comment (gladjohn) [inline msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/ManagedIdentityParameters.java:109]:
```suggestion
        /**
       * Optional claims in JSON format to be included in the token request.
       * Example: {"access_token":{"nbf":{"essential":true, "value":"1743543930"}}}
       *
       * @param claims a valid JSON string representing additional claims
       * @return this builder instance
       */
        public ManagedIdentityParametersBuilder claims(String claims) {
```

File: msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/ManagedIdentityParameters.java [hunk 5]
```diff
@@ -83,8 +86,13 @@ public ManagedIdentityParametersBuilder forceRefresh(boolean forceRefresh) {
...
+            this.claims = claims;
```
Comment (gladjohn) [inline msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/ManagedIdentityParameters.java:112]: we should consider clarifying in code or Javadoc what happens if claims is an empty string, malformed JSON, or null.

File: msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/ManagedIdentityParameters.java [hunk 6]
```diff
@@ -27,7 +29,7 @@ public Set<String> scopes() {
 
     @Override
     public ClaimsRequest claims() {
-        return null;
+        return (claims != null) ? ClaimsRequest.formatAsClaimsRequest(claims) : null;
```
Comment (gladjohn) [inline msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/ManagedIdentityParameters.java]: would logging be helpful here. i.e. claims are present? not the actual claims itself

File: msal4j-sdk/src/test/java/com/microsoft/aad/msal4j/ManagedIdentityTests.java [hunk 7]
```diff
@@ -579,6 +580,38 @@ void managedIdentity_SharedCache(ManagedIdentitySourceType source, String endpoi
         verify(httpClientMock, times(1)).send(any());
     }
 
+    // managedIdentityTest_WithClaims: Tests that acquiring a token with claims works correctly
```
Comment (gladjohn) [inline msal4j-sdk/src/test/java/com/microsoft/aad/msal4j/ManagedIdentityTests.java:621]: The new test method managedIdentityTest_WithClaims is a great start. We might also want to check:

- **Multiple capabilities:** If we pass multiple items in clientCapabilities, do we confirm they are all being sent in the request?

- **Claims edge cases:** If we pass an empty or invalid JSON, does the code handle it gracefully?

File: msal4j-sdk/src/test/java/com/microsoft/aad/msal4j/ManagedIdentityTests.java [hunk 8]
```diff
@@ -579,6 +580,38 @@ void managedIdentity_SharedCache(ManagedIdentitySourceType source, String endpoi
...
+                .httpClient(httpClientMock)
+                .build();
+
+        // Clear caching to avoid cross test pollution.
+        miApp.tokenCache().accessTokens.clear();
+
+        String claimsJson = "{\"default\":\"claim\"}";
+        IAuthenticationResult result = miApp.acquireTokenForManagedIdentity(
+            ManagedIdentityParameters.builder(resource)
+                .claims(claimsJson)
```
Comment (gladjohn) [inline msal4j-sdk/src/test/java/com/microsoft/aad/msal4j/ManagedIdentityTests.java]: do we already have tests that checks the final JSON structure explicitly when both claims and capabilities are set? if not we should consider adding it

File: msal4j-sdk/src/test/java/com/microsoft/aad/msal4j/ManagedIdentityTests.java [hunk 9]
```diff
@@ -579,6 +580,38 @@ void managedIdentity_SharedCache(ManagedIdentitySourceType source, String endpoi
...
+        DefaultHttpClient httpClientMock = mock(DefaultHttpClient.class);
+        if (source == SERVICE_FABRIC) {
+            ServiceFabricManagedIdentitySource.setHttpClient(httpClientMock);
+        }
+
+        when(httpClientMock.send(expectedRequest(source, resource))).thenReturn(expectedResponse(200, getSuccessfulResponse(resource)));
+
+        miApp = ManagedIdentityApplication
+                .builder(ManagedIdentityId.systemAssigned())
+                .clientCapabilities(singletonList("cp1"))
```
Comment (gladjohn) [inline msal4j-sdk/src/test/java/com/microsoft/aad/msal4j/ManagedIdentityTests.java]: would be good to test `null` and empty string `""`

File: msal4j-sdk/src/test/java/com/microsoft/aad/msal4j/ManagedIdentityTests.java [hunk 10]
```diff
@@ -579,6 +580,38 @@ void managedIdentity_SharedCache(ManagedIdentitySourceType source, String endpoi
         verify(httpClientMock, times(1)).send(any());
     }
 
+    // managedIdentityTest_WithClaims: Tests that acquiring a token with claims works correctly
+    @ParameterizedTest
+    @MethodSource("com.microsoft.aad.msal4j.ManagedIdentityTestDataProvider#createDataError")
+    void managedIdentityTest_WithClaims(ManagedIdentitySourceType source, String endpoint) throws Exception {
```
Comment (gladjohn) [inline msal4j-sdk/src/test/java/com/microsoft/aad/msal4j/ManagedIdentityTests.java:624]: The test suggests just claims but even capability is being used.  maybe we can consider changing the test name

File: msal4j-sdk/src/test/java/com/microsoft/aad/msal4j/ManagedIdentityTests.java (hunk 9 above)
Comment (gladjohn) [inline msal4j-sdk/src/test/java/com/microsoft/aad/msal4j/ManagedIdentityTests.java]: we should also consider adding a test without capability. i.e. just use the claims api

File: msal4j-sdk/src/test/java/com/microsoft/aad/msal4j/ManagedIdentityTests.java [hunk 11]
```diff
@@ -579,6 +580,38 @@ void managedIdentity_SharedCache(ManagedIdentitySourceType source, String endpoi
...
+        // Clear caching to avoid cross test pollution.
+        miApp.tokenCache().accessTokens.clear();
+
+        String claimsJson = "{\"default\":\"claim\"}";
+        IAuthenticationResult result = miApp.acquireTokenForManagedIdentity(
+            ManagedIdentityParameters.builder(resource)
+                .claims(claimsJson)
+                .build()).get();
+
+        assertNotNull(result.accessToken());
```
Comment (gladjohn) [inline msal4j-sdk/src/test/java/com/microsoft/aad/msal4j/ManagedIdentityTests.java:649]: are we able to check if the token source is IDP when claims are passed?

File: msal4j-sdk/src/test/java/com/microsoft/aad/msal4j/ManagedIdentityTests.java (hunk 10 above)
Comment (gladjohn) [inline msal4j-sdk/src/test/java/com/microsoft/aad/msal4j/ManagedIdentityTests.java:624]: praise: Excellent new test coverage for the claims JSON scenario!

**Comment by gladjohn** [comment]
The new features for client capabilities and claims looks great. Adding tests for edge cases (like empty or malformed JSON) will strengthen the code coverage further. Excellent job!!!

**Comment by gladjohn** [comment]
nit: would be also good to have a tracking item in the PR description.

File: msal4j-sdk/src/test/java/com/microsoft/aad/msal4j/ManagedIdentityTests.java (hunk 9 above)
Comment (gladjohn) [reply msal4j-sdk/src/test/java/com/microsoft/aad/msal4j/ManagedIdentityTests.java]: reason: app developers can pass claims alone. they need not register for capability, and they should still get a new token.

File: msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/ManagedIdentityApplication.java (hunk 2 above)
Comment (neha-bhargava) [reply msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/ManagedIdentityApplication.java]: Not applying this change since this is consistent with rest of the comments

File: msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/ManagedIdentityParameters.java (hunk 5 above)
Comment (neha-bhargava) [reply msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/ManagedIdentityParameters.java:112]: Is this the format in which the claims are expected? `{"accessToken":{"nbf":{"essential":true}}}`

File: msal4j-sdk/src/test/java/com/microsoft/aad/msal4j/ManagedIdentityTests.java (hunk 7 above)
Comment (neha-bhargava) [reply msal4j-sdk/src/test/java/com/microsoft/aad/msal4j/ManagedIdentityTests.java:621]: In this PR I am only exposing the APIs. The rest of the work will be in another PR. CC: @bgavrilMS

File: msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/ManagedIdentityParameters.java (hunk 5 above)
Comment (gladjohn) [reply msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/ManagedIdentityParameters.java:112]: it will be like this `{"access_token":{"nbf":{"essential":true, "value":"1743543930"}}}`

File: msal4j-sdk/src/test/java/com/microsoft/aad/msal4j/ManagedIdentityTests.java (hunk 8 above)
Comment (neha-bhargava) [reply msal4j-sdk/src/test/java/com/microsoft/aad/msal4j/ManagedIdentityTests.java]: I am adding the APIs in this PR. I am not yet passing this to the endpoints

File: msal4j-sdk/src/test/java/com/microsoft/aad/msal4j/ManagedIdentityTests.java (hunk 7 above)
Comment (gladjohn) [reply msal4j-sdk/src/test/java/com/microsoft/aad/msal4j/ManagedIdentityTests.java:621]: > In this PR I am only exposing the APIs. The rest of the work will be in another PR. CC: @bgavrilMS

**issue (blocking):** Exposing a public API without the corresponding functionality can be risky.

**discussion:** If this API is merged now, consumers might assume it’s ready. We typically prefer to finish the feature (and its tests) in a private or feature branch, then merge once the functionality is complete. Otherwise, we risk confusion or partial breakage for downstream users.

File: msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/ManagedIdentityApplication.java [hunk 12]
```diff
@@ -99,9 +102,21 @@ public Builder resource(String resource) {
...
+         * 
+         * @param clientCapabilities List of client capabilities to be requested
+         * @return instance of Builder of ManagedIdentityApplication
+         */
+        public Builder clientCapabilities(List<String> clientCapabilities) {
+            if (clientCapabilities != null) {
```
Comment (Avery-Dunn) [inline msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/ManagedIdentityApplication.java]: nit: Since the builder's `clientCapabilities` isn't set to a default value it'll be null anyway, so a null check here isn't necessary but also won't cause any problems.

**Review by gladjohn — CHANGES_REQUESTED** [review]
when claims are present, we need to bypass MSALs cache

File: msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/ManagedIdentityApplication.java (hunk 12 above)
Comment (neha-bhargava) [reply msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/ManagedIdentityApplication.java]: Yup, AI generated code I missed to look at.

File: msal4j-sdk/src/test/java/com/microsoft/aad/msal4j/ManagedIdentityTests.java (hunk 7 above)
Comment (bgavrilMS) [reply msal4j-sdk/src/test/java/com/microsoft/aad/msal4j/ManagedIdentityTests.java:621]: Didn't we add the APIs to MSAL.NET in the exact same way @gladjohn ?

https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/c776ab9a8e263787c535463b9518d1521fbd59a9/src/client/Microsoft.Identity.Client/ApiConfig/AcquireTokenForManagedIdentityParameterBuilder.cs#L74

At a minimum, we should bypass MSAL's cache. This is still a value add, even  though we cannot bypass MSI caches.

And add a comment that the APIs will not work (i.e. no-op) on some sources.

Comment (Avery-Dunn) [reply msal4j-sdk/src/test/java/com/microsoft/aad/msal4j/ManagedIdentityTests.java:621]: I don't think merging the APIs without the functionality is a big problem since the APIs related to managed identity won't be used as much as public/confidential, some comments like Bogdan suggested should be enough to let customers know what the situation is.

We could also keep it separate until fully implemented by merging this PR and follow ups into their own branch instead of dev, like I'm doing with some of the partial dependency removals: https://github.com/AzureAD/microsoft-authentication-library-for-java/pull/927

Comment (gladjohn) [reply msal4j-sdk/src/test/java/com/microsoft/aad/msal4j/ManagedIdentityTests.java:621]: > Didn't we add the APIs to MSAL.NET in the exact same way @gladjohn ?

Yes, but we also have added the functionality to bypass cache. 

> At a minimum, we should bypass MSAL's cache. This is still a value add, even though we cannot bypass MSI caches.

agree @bgavrilMS 

@Avery-Dunn – I’d strongly recommend we bypass MSAL’s cache whenever claims are present.

**Reasoning:**
If token revocation occurs for a system-assigned or user-assigned managed identity, bypassing the cache helps us avoid serving a stale token. Instead, we always fetch a new token from the Managed Identity Resource Provider (MIRP). Even if the provider doesn’t fully support token revocation parameters, they can shorten token lifespans or handle partial revocation flows. That way, we reduce the risk of inadvertently returning invalid or expired tokens.

**Next Steps**

- Bypass MSAL’s cache in these scenarios.

This strikes a balance between making the APIs available and ensuring we don’t break consumer expectations regarding token freshness.

File: msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/ManagedIdentityParameters.java (hunk 6 above)
Comment (neha-bhargava) [reply msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/ManagedIdentityParameters.java]: I will throw here instead.

File: msal4j-sdk/src/test/java/com/microsoft/aad/msal4j/ManagedIdentityTests.java (hunk 7 above)
Comment (neha-bhargava) [reply msal4j-sdk/src/test/java/com/microsoft/aad/msal4j/ManagedIdentityTests.java:621]: I will add code to bypass cache when claims are present. That is also something that Bogdan suggested. Thanks everyone for your suggestions and perspectives. My goal here is to keep PRs short and since the whole token revocation with App service new api version cannot be merged, have a PR that can be merged and a small PR with changes that cannot be merged. Short PRs are easy to review and expedites the process. 

Like @Avery-Dunn mentioned can be done as a feature branch as well instead of dev.

File: msal4j-sdk/src/test/java/com/microsoft/aad/msal4j/ManagedIdentityTests.java (hunk 9 above)
Comment (neha-bhargava) [reply msal4j-sdk/src/test/java/com/microsoft/aad/msal4j/ManagedIdentityTests.java]: I will add more tests. Thanks

File: msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/ManagedIdentityApplication.java (hunk 3 above)
Comment (bgavrilMS) [reply msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/ManagedIdentityApplication.java]: I find Neha's comment simpler to read.

File: msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/ManagedIdentityParameters.java (hunk 4 above)
Comment (bgavrilMS) [reply msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/ManagedIdentityParameters.java:109]: The claims string is opaque to the client. We should not give examples, particularly we should not suggest that json is the format.

File: msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/ManagedIdentityParameters.java (hunk 5 above)
Comment (bgavrilMS) [reply msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/ManagedIdentityParameters.java:112]: See above comment - claims are opaque concept. It's a string. Do not perform any validation.

File: msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/ManagedIdentityParameters.java (hunk 4 above)
Comment (bgavrilMS) [reply msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/ManagedIdentityParameters.java:109]: I suggest all MSALs add a comment like:

"Instructs the SDK to bypass any token caches and to request new tokens with an additional claims challenge. The claims challenge string is opaque to applications and should not be parsed. The claims challenge string is issued either by the STS as part of an error response or by the resource, as part of an HTTP 401 response, in the WWW-Authenticate header. For more details see https://learn.microsoft.com/entra/identity-platform/app-resilience-continuous-access-evaluation?tabs=dotnet"

and for CC

"Informs the token issuer that the application is able to perform complex authentication actions. For example, "cp1" means that the application is able to perform conditional access evaluation, because the application has been setup to parse WWW-Authenticate headers associated with a 401 response from the protected APIs, and to retry the request with claims API."

File: msal4j-sdk/src/test/java/com/microsoft/aad/msal4j/ManagedIdentityTests.java (hunk 9 above)
Comment (bgavrilMS) [reply msal4j-sdk/src/test/java/com/microsoft/aad/msal4j/ManagedIdentityTests.java]: @gladjohn - did you document all the test cases developers are required to write? Is there a test plan associated with the spec?

Comment (gladjohn) [reply msal4j-sdk/src/test/java/com/microsoft/aad/msal4j/ManagedIdentityTests.java]: We do not have test case document, but [MSAL .NET unit tests](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/main/tests/Microsoft.Identity.Test.Unit/ManagedIdentityTests/ManagedIdentityTests.cs#L298) should be a good place to start.

File: msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/ManagedIdentityParameters.java (hunk 4 above)
Comment (neha-bhargava) [reply msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/ManagedIdentityParameters.java:109]: Updated

File: msal4j-sdk/src/test/java/com/microsoft/aad/msal4j/ManagedIdentityTests.java (hunk 10 above)
Comment (neha-bhargava) [reply msal4j-sdk/src/test/java/com/microsoft/aad/msal4j/ManagedIdentityTests.java:624]: Updated tests.

**Review by gladjohn — DISMISSED** [review]

File: msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/ManagedIdentityParameters.java [hunk 13]
```diff
@@ -27,7 +29,18 @@ public Set<String> scopes() {
 
     @Override
     public ClaimsRequest claims() {
-        return null;
+        if (claims == null || claims.isEmpty()) {
```
Comment (Avery-Dunn) [inline msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/ManagedIdentityParameters.java:32]: suggestion: If the null/empty check was move to the builder, then the rest of the library could just ignore it and wouldn't have to throw an exception

Comment (neha-bhargava) [reply msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/ManagedIdentityParameters.java:32]: Yeah noticed it when the tests started failing. Always takes a bit to switch to Java style from .Net

**Review by Avery-Dunn — APPROVED** [review]

File: msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/ManagedIdentityParameters.java (hunk 4 above)
Comment (rayluo) [reply msal4j-sdk/src/main/java/com/microsoft/aad/msal4j/ManagedIdentityParameters.java:109]: > I suggest all MSALs add a comment like:
> 
> For MI
>
> "Instructs the SDK to bypass any token caches and to request new tokens with an additional claims challenge. The claims challenge string is opaque to applications and should not be parsed. The claims challenge string is issued either by the STS as part of an error response or by the resource, as part of an HTTP 401 response, in the WWW-Authenticate header. For more details see https://learn.microsoft.com/entra/identity-platform/app-resilience-continuous-access-evaluation?tabs=dotnet"
> 
> and for CC
> 
> "Informs the token issuer that the application is able to perform complex authentication actions. For example, "cp1" means that the application is able to perform conditional access evaluation, because the application has been setup to parse WWW-Authenticate headers associated with a 401 response from the protected APIs, and to retry the request with claims API."

Speaking of claims challenge format, I remember that back then when MSALs were developing that feature for CCA, there was no end-to-end test environment. So, we implemented based on the specs, [which seems to be expecting a string representation of a json blob without base64 encoding](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/4.70.2/src/client/Microsoft.Identity.Client/Internal/ClaimsHelper.cs#L41-L48). Will it work with the base64 encoded sample in [the quoted document](https://learn.microsoft.com/en-us/entra/identity-platform/app-resilience-continuous-access-evaluation?tabs=dotnet#handling-cae-in-your-application)?

Ideally, we shall document the expected claims format accordingly, even if our current implementation in MSI does not (yet?) parse the claim. Otherwise, when/if we decide to parse it in the future, a wrong format input would crash and become a breaking change.
